File: src/leetCode/java/Solution189.java
package leetCode.java;

//O(n) O(1)
class Solution189 {
    public void rotate(int[] nums, int k) {
        if(k % nums.length == 0) return;//無變化

        int start = 0, cur = nums[0], index = 0;
        for(int i=0; i<nums.length; i++){
            index = (index + k) % nums.length;
            int temp = nums[index];
            nums[index] = cur;
            cur = temp;

            if(index == start){//回到原點時
                index = ++start;
                cur = nums[index];
            }
        }
    }
}

File: c++/Solution189.cpp
//O(n) O(1)
class Solution189 {
public:
    void rotate(vector<int>& nums, int k){
        //with O(1) extra space
        if(k % nums.size() == 0) return;//回到原點

        int start = 0, num = nums[0], i = 0;
        for(int j=0; j<nums.size(); ++j){
            i = (i + k) % nums.size();
            int temp = nums[i];
            nums[i] = num;
            num = temp;
            if(i == start){//防止nums.size()為k倍數時回到自己
                i = ++start;
                num = nums[i];
            }
        }
    }
};
/* nums = [1,2,3,4,5,6,7], k = 3
1   2   3   4   5   6   7
            1           4
        7           3
    6           2
5

*/